### ISSUE TITLE
Setting the token.tag_ causes token.lemma_ to reset

### ISSUE THREAD

**Issue by udnaan** [explosion/spaCy#4781]
## How to reproduce the behavior
```
    nlp = spacy.load("en_core_web_sm")
    doc = nlp('Expert kebab maker making kebabs expertly while wearing a funny hat.')

    terms = [[doc[1:3], "kebab baker"]]
    with doc.retokenize() as retokenizer:
        for term in terms:
            retokenizer.merge(term[0], attrs={"LEMMA": term[1]})
    print("Before:", [(token.text, token.lemma_) for token in doc])
    for tok in doc:
        tag = tok.tag_
        tok.tag_ = tag
    print("After:", [(token.text, token.lemma_) for token in doc])


Before: [('Expert', 'expert'), ('kebab maker', 'kebab baker'), ('making', 'make'), ('kebabs', 'kebab'), ('expertly', 'expertly'), ('while', 'while'), ('wearing', 'wear'), ('a', 'a'), ('funny', 'funny'), ('hat', 'hat'), ('.', '.')]
After: [('Expert', 'expert'), ('kebab maker', 'kebab maker'), ('making', 'make'), ('kebabs', 'kebab'), ('expertly', 'expertly'), ('while', 'while'), ('wearing', 'wear'), ('a', 'a'), ('funny', 'funny'), ('hat', 'hat'), ('.', '.')]
```
## Your Environment
OS: Linux (Ubuntu 18.04)
Python: 3.7.4
Spacy: 2.2

## Expected behavior
Changing the tag_ should not affect the custom lemma

## Workaround
Saving lemma and restoring it after setting the tok.tag_ is a workaround.

**Comment by adrianeboyd** [comment]
This is a little tricky, because there are cases where you want the assignment of a different tag to affect the lemma, but I agree that if you're not changing the tag this shouldn't happen.

The technical issue is that's there's a cache that stores lemmas for text+tag combinations, which currently isn't modified when you assign the lemma from the retokenizer (or by hand), so when you change the tag for a text it hasn't seen before, it calculates the lemma again and replaces your custom lemma.

Some options:

1. Modify how lemmas are set so that custom lemmas are added to the cache
  * some potentially unexpected side effects if you have the same orth+tag combo in a future doc, since this value will be in the cache
  * you would need to do this every time an orth is modified, including reassigning the tag to force the lemma lookup at the right point in the in the retokenizer
  * would you want individually assigned lemmas through the lemma property to be added to the cache?
2. Skip tag/lemma assignment if the tag is unchanged
  * if you do tags X -> Y -> X your custom lemma from the original X will be lost

I implemented option 1 to test, but now that I look at the side effects, I think option 2 would be better and simpler.

**Comment by udnaan** [comment]
Take your time to think it through. If nothing else, documenting this and letting the user decide how to handle this should be sufficient.

Additionally, I discovered that pos has similar behavior. Again, the same workaround applies.

**Comment by adrianeboyd** [comment]
A lot of the details related to how lemmas (and pos to some extent) are assigned is pretty hard to follow and more documentation definitely is a good idea.

**Comment by vinodhinir** [comment]
Is this resolved, I have the same problem

**Comment by udnaan** [comment]
@vinodhinir  The workaround I posted still works. Have you tried it?

**Comment by svlandeg** [comment]
It looks like this has now been fixed in spaCy v3:
```
Before: [('Expert', 'expert'), ('kebab maker', 'kebab baker'), ('making', 'make'), ('kebabs', 'kebab'), ('expertly', 'expertly'), ('while', 'while'), ('wearing', 'wear'), ('a', 'a'), ('funny', 'funny'), ('hat', 'hat'), ('.', '.')]
After : [('Expert', 'expert'), ('kebab maker', 'kebab baker'), ('making', 'make'), ('kebabs', 'kebab'), ('expertly', 'expertly'), ('while', 'while'), ('wearing', 'wear'), ('a', 'a'), ('funny', 'funny'), ('hat', 'hat'), ('.', '.')]
```
so I'll go ahead and close this :-)

**Comment by udnaan** [comment]
Awesome. Thanks for the update!
